src/utils/generateData.ts

```diff
@@ -0,0 +1,153 @@
+  const maxDurationMs = 10 * 24 * 60 * 60 * 1000
+  const durationMs = Math.min(rangeInMs, generateRandomNumber(0, maxDurationMs))
+  const randomDate = luxonStartDate.plus({ milliseconds: durationMs })
+  return randomDate.toJSDate().toString()
+}
+
+const generateBatchId = (year: string, batch: Batch[]): string => {
+  const yearsFound = batch.filter((b) => b.year_of_receipt === year)
+  return (yearsFound.length + 1).toLocaleString('en-US', {
+    minimumIntegerDigits: 2,
+    useGrouping: false
+  })
+}
+
+export const generatePlatform = (length: number): Platform[] => {
+  const platforms: Platform[] = []
+  for (let i = 1; i <= length; i++) {
+    platforms.push({ id: i, name: `platform-${i}`, active: true })
+  }
+  return platforms
+}
+
+export const generateProject = (length: number): Project[] => {
+  const projects: Project[] = []
+  for (let i = 1; i <= length; i++) {
+    const [startDate, endDate] = generateRandomDate()
+    const obj: Project = {
+      id: i,
+      name: `project-${i}`,
+      start_date: startDate.toString(),
+      end_date: endDate.toString(),
+      project_code: String(generateRandomNumber(1, 1000)),
+      remarks: `project-remarks-${i}`
+    }
+    projects.push(obj)
+  }
+  return projects
+}
+
+export const generateBatch = (
+  length: number,
+  platforms: number,
+  departments: number,
+  projects: number,
+  organisations: number,
+  protectiveMarkingAuthority: number,
+  protectiveMarking: number
+): Batch[] => {
+  const batches: Batch[] = []
+
+  for (let i = 1; i <= length; i++) {
+    const year = String(generateRandomNumber(2020, 2023))
+    const obj: Batch = {
+      id: i,
+      name: `batch-${i}`,
+      batch_number: `V${generateBatchId(year, batches)}/${year}`,
+      vault: generateRandomNumber(0, length),
+      year_of_receipt: year,
+      department: generateRandomNumber(1, departments - 1),
+      project: generateRandomNumber(1, projects - 1),
+      platform: generateRandomNumber(1, platforms - 1),
+      organisation: generateRandomNumber(1, organisations - 1),
+      protective_marking_authority: generateRandomNumber(
+        1,
+        protectiveMarkingAuthority - 1
+      ),
+      maximum_protective_marking: generateRandomNumber(
+        1,
+        protectiveMarking - 1
+      ),
+      remarks: `remarks-batch-${i}`
+    }
+    batches.push(obj)
+  }
+  return batches
+}
+
+export const generateItems = (
+  length: number,
+  batch: Batch,
+  vaults: number,
+  protectiveMarking: number,
+  project: Project
+): Item[] => {
+  const items: Item[] = []
+  for (let i = 1; i <= length; i++) {
+    const endDate = generateRandomDateInRange(
+      new Date(project.start_date),
+      new Date(project.end_date)
+    )
+    const startDate = generateRandomDateInRange(
+      new Date(project.start_date),
+      new Date(endDate)
+    )
+    const id: number = generateRandomNumber(1, 100000)
+    const obj: Item = {
+      id,
```

Comment:
Hello @TahaKhanAbdalli - for all the objects we create here, we should leave the `id` blank, so that the DataProvider generates it.
We generate random numbers for two of the types, but they may allow duplicate `id` values.
Aah, maybe you are providing the `id` field because TypeScript wants it.

One solution would be to declare a `Partial<Item>`.  But that would allow multiple fields to be missing - and we want TypeScript to force us to provide values for all required fields.  So, instead we could declare a type that [omits](https://stackoverflow.com/a/63991735/92441) the id, and create an instance of that type.

(note: this is right at the edge of my knowledge of TypeScript 🤣 )